lambeq/rewrite/base.py

```diff
@@ -393,3 +422,96 @@ def _ar(self, box: Box) -> Diagram:
 
     def _ob(self, ob: Ty) -> Ty:
         return ob
+
+def dataset_to_words(dataset: Container[str]) -> Container[str]:
+    """ Extracts words from a dataset of sentences.
+    
+    Parameters
+    -----------
+    dataset: Container[str]
+        A list of sentences.
+        
+    Returns
+    -------
+    words: Container[str]
+        A list of words.
+    """
+    dataset_string = ' '.join(dataset)
+    words = dataset_string.split()
+    return words
+    
+class HandleUnknownWords():
+    
+    def __init__(self, training_data: Container[str], test_data: Container[str]):
+        """ Instantiate a handler for unknown words.
+        
+        Parameters
+        ----------
+        training_data: Container[str]
+            List of training dataset sentences
+        test_data: Constainer[str]
+            List of test dataset sentences
+        train_unknown_words: Container[str]
+            List of training dataset words
+        test_unknown_words: Container[str]
+            List of test dataset words
+        train_check: bool
+            Flag for making sure handle_train is run before handle_test
+        """
+        self.training_data = dataset_to_words(training_data)
+        self.test_data = dataset_to_words(test_data)
+        self.train_unknown_words = []
+        self.test_unknown_words = []
+        self.train_check = False
+        
+    def handle_train(self, min_freq: int, diagrams: Container[Diagram]) -> Container[Diagram]:
+        """ Handles training dataset unknown words, and tokenizes the training diagrams.
+        
+        Parameters
+        ----------
+        min_freq: int
+            The minimum frequency threshold for indicating unknown words
+        diagrams: Container[Diagram]
+            The list of training dataset parsed diagrams to be tokenized
+            
+        Returns
+        -------
+        diagrams: Container[Diagram] 
+        """
+        self.train_check = True
+        words = [i for i in self.training_data if self.training_data.count(i) < min_freq]
+        self.train_unknown_words = words
+        
+        rewriter = Rewriter([UNKRewriteRule(words = words)])
+        diagrams = [rewriter(i) for i in diagrams]
+        
+        return diagrams
+    
+    def handle_test(self) -> Container[Diagram]:
```

Comment:
The diagrams are not passed here, so line 502 will fail below.

Reply:
I see! I'll fix it right away!